examples/Custom_Hooks/probability_of_improvement.py

```diff
@@ -0,0 +1,219 @@
+## Example for calculating and plotting the Probability of Improvement
+
+# This example demonstrates how to use the
+# {func}`register_hooks <baybe.utils.basic.register_hooks>` utility
+# to calculate and analyse the probability of improvement (PI) evolving during iterations:
+# * We define a hook that is compatible with the general
+#   {meth}`RecommenderProtocol.recommend <baybe.recommenders.base.RecommenderProtocol.recommend>`
+#   interface,
+# * attach it to a recommender,
+# * and watch it take action.
+
+
+### Imports
+
+import os
+import warnings
+from pathlib import Path
+from types import MethodType
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import torch
+from botorch.test_functions.synthetic import Hartmann
+from matplotlib.axes import Axes
+from matplotlib.collections import PolyCollection
+from matplotlib.figure import Figure
+from scipy.stats import gaussian_kde
+
+from baybe.acquisition import ProbabilityOfImprovement
+from baybe.campaign import Campaign
+from baybe.objectives.base import Objective
+from baybe.objectives.single import SingleTargetObjective
+from baybe.parameters import NumericalDiscreteParameter
+from baybe.recommenders import (
+    BotorchRecommender,
+    RandomRecommender,
+    TwoPhaseMetaRecommender,
+)
+from baybe.searchspace import SearchSpace, SearchSpaceType
+from baybe.surrogates import GaussianProcessSurrogate
+from baybe.targets import NumericalTarget
+from baybe.utils.basic import register_hooks
+from baybe.utils.botorch_wrapper import botorch_function_wrapper
+from baybe.utils.dataframe import to_tensor
+
+### Parameters for a full simulation loop
+
+# For the full simulation, we need to define some additional parameters. These are the number of experiments to be conducted per run, the batch size, the dimension and the points per dimension.
+
+SMOKE_TEST = "SMOKE_TEST" in os.environ
+N_DOE_ITERATIONS = 3 if SMOKE_TEST else 7
+BATCH_SIZE = 2
+DIMENSION = 3
+POINTS_PER_DIM = 4
+
+### Setup
+
+# We start by initializing the global `poi_list` where we will store the PI from each recommend iteration:
+
+pi_list: list[torch.Tensor] = []
+
+
+# Then, we define the hook that calculates the PI based on the given recommender, search space and measurements:
+
+
+def check_probability_of_improvement(
+    self: BotorchRecommender,
+    searchspace: SearchSpace,
+    objective: Objective | None = None,
+    measurements: pd.DataFrame | None = None,
+) -> list[torch.Tensor]:
```

Comment:
Functions should always operate on the lowest required level of inputs/outputs. No torch is required here. So better use `np.ndarray`. Also, since the arrays are all of equal length, you can actually just stack them into a single array. Accordingly, you also need to change the type of your `pi_list` variable. Actually, I just notice that the function doesn't return anything at all, so the return type needs to be simply `None`, but the type of your list needs to change

Reply:
not a fan of stacking the results from iterations (if thats what you meant) but keeping it list

Reply:
@AdrianSosic @Scienfitz So do I just change the type to `np.array`?  I tried it and it works.

Reply:
No, of course you also need to cast the output accordingly. @Scienfitz is not a fan of having a single array (which is OK for me), so that means your output would be `list[np.ndarray]` and inside your function you need to call `to_numpy()` in the appropriate place

Reply:
Well, I changed the type to `list[np.array]` but did not need to do extra cast operations, because, I already casted it where it is used (see line 109).

Reply:
* first of all: `list[np.ndarray]` instead of `list[np.array]`
* Yes, you'll get no runtime error (because python is a dynamically typed language) but it's definitely wrong and `mypy` should also throw you an error

Reply:
Is it possible to use `pi.detach().numpy()` instead of `to_numpy()`? I tried this and the mypy error  were solved.

Reply:
Yes, but probably it would be cleaner to embed the `botorch_acqf()` call into a `with torch.no_grad()` context to avoid that the computational graph is constructed in the first place. Then you don't need the `detach` call